File: archive/pulumi-java/Pr3916/lib/src/main/java/app/SecurityStack.java
package app;

import com.pulumi.Context;
import com.pulumi.aws.wafv2.WebAcl;
import com.pulumi.aws.wafv2.WebAclArgs;
import com.pulumi.aws.wafv2.inputs.WebAclDefaultActionArgs;
import com.pulumi.aws.wafv2.inputs.WebAclDefaultActionAllowArgs;
import com.pulumi.aws.wafv2.inputs.WebAclRuleArgs;
import com.pulumi.aws.wafv2.inputs.WebAclRuleActionArgs;
import com.pulumi.aws.wafv2.inputs.WebAclRuleActionBlockArgs;
import com.pulumi.aws.wafv2.inputs.WebAclRuleStatementArgs;
import com.pulumi.aws.wafv2.inputs.WebAclRuleStatementRateBasedStatementArgs;
import com.pulumi.aws.wafv2.inputs.WebAclRuleVisibilityConfigArgs;
import com.pulumi.aws.wafv2.inputs.WebAclRuleOverrideActionArgs;
import com.pulumi.aws.wafv2.inputs.WebAclRuleOverrideActionNoneArgs;
import com.pulumi.aws.wafv2.inputs.WebAclRuleStatementManagedRuleGroupStatementArgs;
import com.pulumi.aws.wafv2.inputs.WebAclVisibilityConfigArgs;

import java.util.Map;

/**
 * Security stack for WAF configuration.
 */
public class SecurityStack {
    private final WebAcl webAcl;

    public SecurityStack(final Context ctx) {
        // Create WAF Web ACL with rate limiting
        this.webAcl = new WebAcl("news-portal-waf",
            WebAclArgs.builder()
                .name("news-portal-waf")
                .description("WAF for news portal with rate limiting")
                .scope("CLOUDFRONT")
                .defaultAction(
                    WebAclDefaultActionArgs.builder()
                        .allow(WebAclDefaultActionAllowArgs.builder().build())
                        .build()
                )
                .rules(
                    // Rate limiting rule - 2000 requests per 5 minutes
                    WebAclRuleArgs.builder()
                        .name("rate-limit-rule")
                        .priority(1)
                        .action(
                            WebAclRuleActionArgs.builder()
                                .block(WebAclRuleActionBlockArgs.builder().build())
                                .build()
                        )
                        .statement(
                            WebAclRuleStatementArgs.builder()
                                .rateBasedStatement(
                                    WebAclRuleStatementRateBasedStatementArgs.builder()
                                        .limit(2000)
                                        .aggregateKeyType("IP")
                                        .build()
                                )
                                .build()
                        )
                        .visibilityConfig(
                            WebAclRuleVisibilityConfigArgs.builder()
                                .cloudwatchMetricsEnabled(true)
                                .metricName("RateLimitRule")
                                .sampledRequestsEnabled(true)
                                .build()
                        )
                        .build(),
                    // AWS Managed Rules - Common Rule Set
                    WebAclRuleArgs.builder()
                        .name("aws-managed-rules")
                        .priority(2)
                        .overrideAction(
                            WebAclRuleOverrideActionArgs.builder()
                                .none(WebAclRuleOverrideActionNoneArgs.builder().build())
                                .build()
                        )
                        .statement(
                            WebAclRuleStatementArgs.builder()
                                .managedRuleGroupStatement(
                                    WebAclRuleStatementManagedRuleGroupStatementArgs.builder()
                                        .vendorName("AWS")
                                        .name("AWSManagedRulesCommonRuleSet")
                                        .build()
                                )
                                .build()
                        )
                        .visibilityConfig(
                            WebAclRuleVisibilityConfigArgs.builder()
                                .cloudwatchMetricsEnabled(true)
                                .metricName("AWSManagedRules")
                                .sampledRequestsEnabled(true)
                                .build()
                        )
                        .build()
                )
                .visibilityConfig(
                    WebAclVisibilityConfigArgs.builder()
                        .cloudwatchMetricsEnabled(true)
                        .metricName("NewsPortalWAF")
                        .sampledRequestsEnabled(true)
                        .build()
                )
                .tags(Map.of(
                    "Name", "NewsPortalWAF",
                    "Environment", "production"
                ))
                .build());

        // Note: WAF association with CloudFront is handled in the distribution configuration
    }

    public WebAcl getWebAcl() {
        return webAcl;
    }
}

File: archive/cdk-ts/Pr4560/lib/security-stack.ts
import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as wafv2 from 'aws-cdk-lib/aws-wafv2';

export interface SecurityStackProps extends cdk.StackProps {
  environmentSuffix: string;
}

export class SecurityStack extends cdk.Stack {
  public readonly webAcl: wafv2.CfnWebACL;

  constructor(scope: Construct, id: string, props: SecurityStackProps) {
    super(scope, id, props);

    // Create WAF Web ACL for CloudFront (must be in us-east-1 for CloudFront)
    this.webAcl = new wafv2.CfnWebACL(
      this,
      `PaymentsWebAcl-${props.environmentSuffix}`,
      {
        name: `payments-gateway-web-acl-${props.environmentSuffix}`,
        defaultAction: { allow: {} },
        scope: 'CLOUDFRONT',
        visibilityConfig: {
          cloudWatchMetricsEnabled: true,
          sampledRequestsEnabled: true,
          metricName: `PaymentsWebAcl-${props.environmentSuffix}`,
        },
        rules: [
          // AWS Managed Rules Core rule set
          {
            name: 'AWSManagedRulesCommonRuleSet',
            priority: 0,
            statement: {
              managedRuleGroupStatement: {
                vendorName: 'AWS',
                name: 'AWSManagedRulesCommonRuleSet',
              },
            },
            overrideAction: { none: {} },
            visibilityConfig: {
              cloudWatchMetricsEnabled: true,
              sampledRequestsEnabled: true,
              metricName: 'AWSManagedRulesCommonRuleSet',
            },
          },
          // AWS Managed Rules IP reputation list
          {
            name: 'AWSManagedRulesAmazonIpReputationList',
            priority: 1,
            statement: {
              managedRuleGroupStatement: {
                vendorName: 'AWS',
                name: 'AWSManagedRulesAmazonIpReputationList',
              },
            },
            overrideAction: { none: {} },
            visibilityConfig: {
              cloudWatchMetricsEnabled: true,
              sampledRequestsEnabled: true,
              metricName: 'AWSManagedRulesAmazonIpReputationList',
            },
          },
          // Rate limiting rule to protect from DDoS
          {
            name: 'RateLimitRule',
            priority: 2,
            statement: {
              rateBasedStatement: {
                limit: 2000,
                aggregateKeyType: 'IP',
              },
            },
            action: { block: {} },
            visibilityConfig: {
              cloudWatchMetricsEnabled: true,
              sampledRequestsEnabled: true,
              metricName: 'RateLimitRule',
            },
          },
        ],
      }
    );

    // Note: WebACL is associated with CloudFront distribution during distribution creation
    // See GlobalStack where webAclId is set on the Distribution

    // Output values for integration testing
    new cdk.CfnOutput(this, 'WebAclArn', {
      value: this.webAcl.attrArn,
      description: 'ARN of the WAF Web ACL',
      exportName: `WebAclArn-${props.environmentSuffix}`,
    });

    new cdk.CfnOutput(this, 'WebAclId', {
      value: this.webAcl.attrId,
      description: 'ID of the WAF Web ACL',
      exportName: `WebAclId-${props.environmentSuffix}`,
    });

    new cdk.CfnOutput(this, 'WebAclName', {
      value: `payments-gateway-web-acl-${props.environmentSuffix}`,
      description: 'Name of the WAF Web ACL',
      exportName: `WebAclName-${props.environmentSuffix}`,
    });

    new cdk.CfnOutput(this, 'WebAclScope', {
      value: 'CLOUDFRONT',
      description: 'Scope of the WAF Web ACL',
      exportName: `WebAclScope-${props.environmentSuffix}`,
    });

    new cdk.CfnOutput(this, 'RateLimitRule', {
      value: '2000 requests per 5 minutes per IP',
      description: 'Rate limit configuration',
    });

    new cdk.CfnOutput(this, 'ManagedRules', {
      value:
        'AWSManagedRulesCommonRuleSet, AWSManagedRulesAmazonIpReputationList',
      description: 'Enabled AWS Managed Rules',
    });
  }
}
